internal/plugin/status.go

```diff
@@ -0,0 +1,163 @@
+package plugin
+
+import (
+	"context"
+	"reflect"
+
+	"github.com/golang/protobuf/ptypes/empty"
+	"github.com/hashicorp/go-argmapper"
+	"github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/go-plugin"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"github.com/hashicorp/waypoint-plugin-sdk/component"
+	"github.com/hashicorp/waypoint-plugin-sdk/internal/funcspec"
+	"github.com/hashicorp/waypoint-plugin-sdk/internal/pluginargs"
+	pb "github.com/hashicorp/waypoint-plugin-sdk/proto/gen"
+)
+
+// statusClient implements component.Status for a service that
+// has the status methods implemented.
+type statusClient struct {
+	Client  statusProtoClient
+	Logger  hclog.Logger
+	Broker  *plugin.GRPCBroker
+	Mappers []*argmapper.Func
+}
+
+func (c *statusClient) Implements(ctx context.Context) (bool, error) {
+	if c == nil {
+		return false, nil
+	}
+
+	resp, err := c.Client.IsStatus(ctx, &empty.Empty{})
+	if err != nil {
+		return false, err
+	}
+
+	return resp.Implements, nil
+}
+
+func (c *statusClient) StatusFunc() interface{} {
+	impl, err := c.Implements(context.Background())
+	if err != nil {
+		return funcErr(err)
+	}
+	if !impl {
+		return nil
+	}
+
+	// Get the spec
+	spec, err := c.Client.StatusSpec(context.Background(), &empty.Empty{})
+	if err != nil {
+		return funcErr(err)
+	}
+
+	spec.Result = nil
+
+	return funcspec.Func(spec, c.status,
+		argmapper.Logger(c.Logger),
+		argmapper.Typed(&pluginargs.Internal{
+			Broker:  c.Broker,
+			Mappers: c.Mappers,
+			Cleanup: &pluginargs.Cleanup{},
+		}),
+	)
+}
+
+func (c *statusClient) status(
+	ctx context.Context,
+	args funcspec.Args,
+	internal *pluginargs.Internal,
+) (*pb.StatusReport, error) {
+	// Run the cleanup
+	defer internal.Cleanup.Close()
+
+	// Call our function
+	resp, err := c.Client.Status(ctx, &pb.FuncSpec_Args{Args: args})
+	if err != nil {
+		return nil, err
+	}
+
+	// just return `resp` here, pass it along
+	return &pb.StatusReport{
+		Resources:     resp.Resources,
+		External:      resp.External,
+		HealthMessage: resp.HealthMessage,
+		TimeGenerated: resp.TimeGenerated,
+		Health:        resp.Health,
+	}, nil
```

Comment:
Is there a reason you're recreating the struct here? resp is already a pb.StatusReport, right?

Reply:
Correct....I think this reconstruction here was left over from when I was trying to have a StatusReport struct for callers to use but settled on having them just use the proto. I'll simplify this!